### PR TITLE
Adding support for variable allocated textures to the GL 4.1 backend

### PR DESCRIPTION
**Pull request by jherico**
This PR adds the same dynamic re-allocation functionality currently used for the GL 4.5 backend to the GL 4.1 backend.  A significant amount of the code for managing the variable allocated textures has been moved to a common support class.  

## Testing

Testing should be done using the [Variable Allocated Textures](https://highfidelity.atlassian.net/wiki/display/HOME/Variable+Allocated+Textures) test plan in conjunction with the [instructions](https://highfidelity.atlassian.net/wiki/display/HOME/Testing+OpenGL+4.1+functionality) for forcing interface to use the OpenGL 4.1 backend.  

The application should also be tested in the same way without forcing the OpenGL 4.1 backend, as some code for the OpenGL 4.5 implementation has moved and is now shared by both implementations.  

## OpenGL 4.1 testing

* Set an environment variable on your computer with the name HIFI_DISABLE_OPENGL_45.  The value of the environment variable does not matter, so long as it's set.  Once you've set this variable, you can start Interface
* Open Interface
* Verify that you are using the 4.1 back-end by opening up the log file viewer (Ctrl-Shift-L) and entering the text Using OpenGL in the search field.  If you have successfully set the environment variable then you should see a line that says `Using OpenGL 4.1 backend`
* Perform the testing for variable allocated textures 
* Make sure when you are finished testing to remove the HIFI_DISABLE_OPENGL_45 environment variable, or it could negatively impact your performance on a system that would otherwise be using the OpenGL 4.5 backend

## Variable Allocated Textures Testing

For testing it is desirable to have the stats display (which can toggled with the '/' key) turned on, as it includes a real time display of the currently consumed texture memory (listed as `Resource Memory`) as well as the amount of texture information which is waiting to be populated (listed as `Pending Transfer`).  When testing it's best to wait until the current state is stable and the Pending Transfer value is zero before making observations.  
To test the variable allocated texture functionality...

### Setup

* Start Interface
* Ensure that the current allowed memory is set to it's default value by going into `Developer | Render | Maximum Texture Memory` menu and ensuring that Automatic Texture Memory is selected
* Go to a domain with lots of content that includes many textures.  Typically `Playa or `Dev-Playa` are ideal as they consume a significant amount of texture memory.
* Spin the viewpoint around a couple of times to ensure that most of the domain content is loaded into memory
* View the stats display and wait for the `Pending Transfer` value to fall to zero.  The `Resource Memory` value should be at least 512 MB.
* Position yourself near one of the models in the scene with a high quality texture, such as directly in front of the Circus Casino sign

### Over-subscription testing

* Open the `Developer | Render | Maximum Texture Memory` menu and select 64 MB
* The `Resource Memory` value should rapidly fall to less than 64 MB.  Concurrently the quality of the textures in the rendered scene should visibly drop

### Under-subscription testing

* Open the `Developer | Render | Maximum Texture Memory` menu and reset it to Automatic Texture Memory
* The `Resource Memory` value should rapidly rise to it's previous value, and the `Pending Transfer` value should jump up to nearly that amount as well.
* The `Pending Transfer` value should then start to fall.  As it does so, the textured items in the scene should return to their previous quality.

### CODE REVIEW

**Comment by ZappoMan** [comment]
@jherico - this is great. Please go add some test plans to the test plan library that include using the "allowed texture memory" setting to verify quality changing. Please include tests that explain the use of the non 4.5 backend on windows, and how to verify that it's working.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 359a7a039e50c19258674f4a95c6942c519e584f available at: https://gist.github.com/908d5b90f20c1f78de21d5e69bff0719<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10208-359a7a0.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10208-359a7a0.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b2aa3271f75ff33644db7e2ea54d69114ad05d32 available at: https://gist.github.com/908d5b90f20c1f78de21d5e69bff0719<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10208-b2aa327.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10208-b2aa327.dmg

**Comment by Nex-Pro** [comment]
@jherico @ZappoMan We don't have access AFAIK to the atlassian thing https://highfidelity.atlassian.net/wiki/display/HOME/Variable+Allocated+Textures

File: libraries/gpu-gl/src/gpu/gl/GLBackend.cpp [hunk 1]
```diff
@@ -744,6 +744,10 @@ void GLBackend::recycle() const {
             glDeleteQueries((GLsizei)ids.size(), ids.data());
         }
     }
+
+    GLVariableAllocationSupport::manageMemory();
+    GLVariableAllocationSupport::_frameTexturesCreated = 0;
```
Comment (samcake) [inline libraries/gpu-gl/src/gpu/gl/GLBackend.cpp]: Maybe worth a function like "resetFrameCounters" or something like that no ?
Not a show  stopper

**Comment by OmegaHeron** [comment]
As @nex-pro mentioned - will need a copy/paste of info into this before QA can proceed.

**Comment by OmegaHeron** [comment]
Windows 10 - i5 6500 - GTX970
QA Fail

Once texture RAM is set to low value for this PR, regardless of GL4.1 or GL4.5  - it never stabilizes, there's a non ending rise/fall to pending transfer and resource memory.  I ran same test with current master (6336) and saw nothing like what I'm seeing with this.

Something seems wrong here in 4.1 Mode - It never seems to stabilize when texture memory is forced to low value (64M).

Location/orientation used:  hifi://dev-playa/-2.42453,-8.09158,-14.9142/0,-0.533467,0,0.845821

This PR GL4.1
Pending Transfer = 0, Resource Memory = 747 In front of circus sign.
Set Texture Memory to 64M - Pending Transfer and Resource memory never stabilize - I see pending transfer drop to 9M then returns to some higher value (600ish).  Resource memory cycles between 600ish MB and 9MB.  This does not stop.

This PR GL4.5
Same as above - endless cycle of Pending transfer and resource memory.

---
Comparison to master

Master-6336 GL4.5
Pending Transfer = 0, Resource Memory = 759 In front of circus sign.
Set Texture memory to 64MB, Resource Memory = 60 and quality dropped as expected.
Set Texture memory to Auto, Resource Memory = 772 and quality returned to former (high) appearance.

Master GL4.1
Pending Transfer = 0, Resource Memory = 775 In front of circus sign.
Set Texture memory to 64MB, Resource Memory = 775 - no change (as expected, I think).
Set Texture memory to Auto, Resource Memory = 775 - no change (as expected, I think).

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2a2102c7fe3c1239acf651121163966fa7e504d1 available at: https://gist.github.com/908d5b90f20c1f78de21d5e69bff0719<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10208-2a2102c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10208-2a2102c.dmg

**Comment by jherico** [comment]
@OmegaHeron thanks, can you try the new build?

**Comment by OmegaHeron** [comment]
Will do @jherico , be a bit before I can.  Will try tonight if possible.

**Comment by OmegaHeron** [comment]
Windows 10 - i5 6500 - GTX970
Seems correct now, but, one *important* question before I say pass.

Should resource memory continue to climb as I start in one domain and move to others?  Just going between dev-playa and heron I see it has climbed from 771 to 3271 after 3 trips.

---

GL4.5 This PR
Pending Transfer = 0, Resource Memory = 771 In front of circus sign.
Set Texture memory to 64MB, Pending Transfer = 0, Resource Memory = 60 and quality dropped as expected.
Set Texture memory to Auto, Pending Transfer = 0, Resource Memory = 771 and quality returned to former (high) appearance.

GL4.1 This PR
Pending Transfer = 0, Resource Memory = 771 In front of circus sign.
Set Texture memory to 64MB, Pending Transfer = 0, Resource Memory = 60 and quality dropped as expected.
Set Texture memory to Auto, Pending Transfer = 0, Resource Memory = 771 and quality returned to former (high) appearance.

**Comment by OmegaHeron** [comment]
Ok - @jherico - I think for sure there's a fail still.

Master or this PR in GL4.5 shows Resource memory coming to some peak value in certain place then dropping and coming to a new value on domain change.  This PR in GL4.1 shows an ever increasing value to Resource Memory as you change domains.

**Comment by Nex-Pro** [comment]
Testing this with i7 Win10 and Nvidia gtx1070

**Comment by Nex-Pro** [comment]
### #10208 QA.

Testing platform: Intel workstation.
OS: Win10 pro 64
CPU: i7 4.0 ghz
RAM: 16gb
GPU: Nvidia GTX1070 8gb VRam
HMD: Htc Vive.

I am getting similar results to @omegaheron. 
Therefore reluctant to pass this on my system. 

Initial testing at dev-playa was all good.
However after I went to another domain ( qbit Venice ) I observed resource mem climb to 1800 MB.

Returned to dev-playa ( same spot ) Observed resource mem climb to 2893 MB.

Back to Venice ... 4004 MB  then dev-playa  5112 is this right ?


Installed client and domain server.

No skybox at domain sandbox. Black sky only in 4.5 and 4.1.

Memory usage seems fairly stable across multiple restarts of interface at same location in dev-playa.


Log output for envVar.
    [04/19 17:06:46] [DEBUG] [hifi.gpu.gl] Using OpenGL 4.1 backend

Start Interface ( restarted after going to dev-playa )

Ensure that the current allowed memory is set to it's default value by going into Developer | Render | Maximum Texture Memory menu and ensuring that Automatic Texture Memory is selected: **CHECK**

Go to a domain with lots of content that includes many textures. 
Typically Playa orDev-Playa` are ideal as they consume a significant amount of texture memory.

Spin the viewpoint around a couple of times to ensure that most of the domain content is loaded into memory
View the stats display and wait for the Pending Transfer value to fall to zero. **DONE**

The Resource Memory value should be at least 512 MB. **584 MB actual**

Position yourself near one of the models in the scene with a high quality texture, such as directly in front of the Circus Casino sign **DONE**


### Over-subscription testing

Open the Developer | Render | Maximum Texture Memory menu and select 64 MB **DONE**
The Resource Memory value should rapidly fall to less than 64 MB. **Dropped to 60 MB**
Concurrently the quality of the textures in the rendered scene should visibly drop **Quality dropped significantly.**


### Under-subscription testing

Open the Developer | Render | Maximum Texture Memory menu and reset it to Automatic Texture Memory **DONE**
The Resource Memory value should rapidly rise to it's previous value, **SUCCESS**
and the Pending Transfer value should jump up to nearly that amount as well. **SUCCESS**
The Pending Transfer value should then start to fall. As it does so, the textured items in the scene should return to their previous quality. **SUCCESS**

**Comment by hifi-gustavo** [comment]
Build console output for commit: bccc06d2fe6f228e44b09934992279f4189c3122 available at: https://gist.github.com/908d5b90f20c1f78de21d5e69bff0719<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10208-bccc06d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10208-bccc06d.dmg

**Comment by samcake** [comment]
@OmegaHeron @Nex-Pro Can you have a second look ?
It sounds like the problem you were repporting is just about the stats beeing displayed, it should be fixed now

**Comment by OmegaHeron** [comment]
Will do.

**Comment by OmegaHeron** [comment]
Windows 10 - i5 6500 - GTX970
QA Pass

Specific tests pass.
GL4.1 now shows same change for resource memory on domain change as 4.5
Traced memory footprint for application comparing to master.  No issue seen.

**Comment by samcake** [comment]
Alright, we re good to go, will merge asap

**Comment by jherico** [comment]
@OmegaHeron great find...  should be fixed.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 472835d65c627285331f7639a2b1e172382afd27 available at: https://gist.github.com/908d5b90f20c1f78de21d5e69bff0719<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10208-472835d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10208-472835d.dmg

**Pull request merged**